File: Dart/13-Heap(PriorityQueue)/Algorithms/1-MinHeap-MaxHeap/MinHeap-MaxHeap.dart
// Min Heap Implementation
class MinHeap {
  // The list to store the heap elements.
  List<int> heap = [];

  // Adds a value to the min heap.
  void add(int val) {
    heap.add(val); // Add the new value to the end of the list.
    _heapifyUp(heap.length - 1); // Maintain heap property by moving the new element up.
  }

  // Removes and returns the minimum value from the heap.
  int removeFirst() {
    if (heap.isEmpty) return -1; // Return -1 if the heap is empty.
    int min = heap.first; // Store the minimum value (root).
    heap[0] = heap.removeLast(); // Replace the root with the last element and remove the last element.
    _heapifyDown(0); // Maintain heap property by moving the new root down.
    return min; // Return the original minimum value.
  }

  // Gets the first element (minimum) without removing it.
  int get first => heap.isNotEmpty ? heap.first : -1;

  // Gets the number of elements in the heap.
  int get length => heap.length;

  // Helper method to maintain min heap property while adding an element.
  void _heapifyUp(int index) {
    while (index > 0) {
      int parent = (index - 1) ~/ 2; // Calculate the parent's index.  ~/: Integer division
      if (heap[parent] <= heap[index]) break; // If parent is smaller or equal, heap property is satisfied.
      _swap(index, parent); // Otherwise, swap the node with its parent.
      index = parent; // Update the index to the parent's index to continue heapifying up.
    }
  }

  // Helper method to maintain min heap property while removing the root.
  void _heapifyDown(int index) {
    int leftChild, rightChild, smallest;
    while (index < heap.length ~/ 2) { // Continue as long as the node has at least one child.
      leftChild = 2 * index + 1; // Calculate the left child's index.
      rightChild = 2 * index + 2; // Calculate the right child's index.
      smallest = leftChild; // Assume the left child is the smallest.
      if (rightChild < heap.length && heap[rightChild] < heap[leftChild]) {
        smallest = rightChild; // If the right child is smaller, update smallest.
      }
      if (heap[index] <= heap[smallest]) break; // If the current node is smaller or equal to the smallest child, heap property is satisfied.
      _swap(index, smallest); // Otherwise, swap the node with the smallest child.
      index = smallest; // Update the index to the smallest child's index to continue heapifying down.
    }
  }

  // Helper method to swap two elements in the heap.
  void _swap(int i, int j) {
    int temp = heap[i];
    heap[i] = heap[j];
    heap[j] = temp;
  }
}

// Max Heap Implementation
class MaxHeap {
  // The list to store the heap elements.
  List<int> heap = [];

  // Adds a value to the max heap.
  void add(int val) {
    heap.add(val); // Add the new value to the end of the list.
    _heapifyUp(heap.length - 1); // Maintain heap property by moving the new element up.
  }

  // Removes and returns the maximum value from the heap.
  int removeFirst() {
    if (heap.isEmpty) return -1; // Return -1 if the heap is empty.
    int max = heap.first; // Store the maximum value (root).
    heap[0] = heap.removeLast(); // Replace the root with the last element and remove the last element.
    _heapifyDown(0); // Maintain heap property by moving the new root down.
    return max; // Return the original maximum value.
  }

  // Gets the first element (maximum) without removing it.
  int get first => heap.isNotEmpty ? heap.first : -1;

  // Gets the number of elements in the heap.
  int get length => heap.length;

  // Helper method to maintain max heap property while adding an element.
  void _heapifyUp(int index) {
    while (index > 0) {
      int parent = (index - 1) ~/ 2; // Calculate the parent's index.
      if (heap[parent] >= heap[index]) break; // If parent is larger or equal, heap property is satisfied.
      _swap(index, parent); // Otherwise, swap the node with its parent.
      index = parent; // Update the index to the parent's index to continue heapifying up.
    }
  }

  // Helper method to maintain max heap property while removing the root.
  void _heapifyDown(int index) {
    int leftChild, rightChild, largest;
    while (index < heap.length ~/ 2) { // Continue as long as the node has at least one child.
      leftChild = 2 * index + 1; // Calculate the left child's index.
      rightChild = 2 * index + 2; // Calculate the right child's index.
      largest = leftChild; // Assume the left child is the largest.
      if (rightChild < heap.length && heap[rightChild] > heap[leftChild]) {
        largest = rightChild; // If the right child is larger, update largest.
      }
      if (heap[index] >= heap[largest]) break; // If the current node is larger or equal to the largest child, heap property is satisfied.
      _swap(index, largest); // Otherwise, swap the node with the largest child.
      index = largest; // Update the index to the largest child's index to continue heapifying down.
    }
  }

  // Helper method to swap two elements in the heap.
  void _swap(int i, int j) {
    int temp = heap[i];
    heap[i] = heap[j];
    heap[j] = temp;
  }
}

// Approach 1: Using Sorting
// Finds the kth largest element by sorting the entire array.
int findKthLargestSorting(List<int> nums, int k) {
  nums.sort(); // Sort the array in ascending order.
  return nums[nums.length - k]; // The kth largest element is at index (length - k).
}

// Approach 2: Using a Min Heap
// Finds the kth largest element using a min heap.
int findKthLargestMinHeap(List<int> nums, int k) {
  MinHeap minHeap = MinHeap(); // Create a new min heap.
  for (var num in nums) {
    minHeap.add(num); // Add each number from the input array to the min heap.
    if (minHeap.length > k) {
      minHeap.removeFirst(); // If the heap size exceeds k, remove the smallest element.
    }
  }
  return minHeap.first; // The root of the min heap will be the kth largest element.
}

// Approach 3: Using a Max Heap
// Finds the kth largest element using a max heap.
int findKthLargestMaxHeap(List<int> nums, int k) {
  MaxHeap maxHeap = MaxHeap(); // Create a new max heap.
  maxHeap.heap.addAll(nums);
  maxHeap.heap.sort((a, b) => b.compareTo(a));
  return maxHeap.heap[k - 1];
}

// Approach 4: Quickselect Algorithm
// Finds the kth largest element using the Quickselect algorithm (efficient).
int quickSelect(List<int> nums, int k) {
  // Partition function to rearrange the array around a pivot.
  int partition(List<int> arr, int left, int right) {
    int pivot = arr[right]; // Choose the rightmost element as the pivot.
    int pIndex = left; // Index to track the position where elements greater than pivot should go.
    for (int i = left; i < right; i++) {
      if (arr[i] >= pivot) { // If the current element is greater than or equal to the pivot.
        int temp = arr[i]; // Swap arr[i] and arr[pIndex].
        arr[i] = arr[pIndex];
        arr[pIndex] = temp;
        pIndex++; // Increment pIndex to point to the next position for a larger element.
      }
    }
    arr[right] = arr[pIndex]; // Place the pivot in its correct position.
    arr[pIndex] = pivot;
    return pIndex; // Return the index of the pivot.
  }

  int left = 0, right = nums.length - 1; // Initialize left and right pointers.
  while (true) {
    int pivotIndex = partition(nums, left, right); // Partition the array.
    if (pivotIndex == k - 1) return nums[pivotIndex]; // If the pivot is at the kth position, we found the element.
    else if (pivotIndex < k - 1) left = pivotIndex + 1; // If pivot is too small, search in the right subarray.
    else right = pivotIndex - 1; // If pivot is too large, search in the left subarray.
  }
}

// Approach 5: Using Counting Sort (for small numbers range)
// Finds the kth largest element using counting sort (efficient for limited range).
int findKthLargestCountingSort(List<int> nums, int k) {
  int maxVal = nums.reduce((a, b) => a > b ? a : b); // Find the maximum value in the array.
  int minVal = nums.reduce((a, b) => a < b ? a : b); // Find the minimum value in the array.
  List<int> count = List.filled(maxVal - minVal + 1, 0); // Create a count array to store the frequency of each number.

  for (var num in nums) {
    count[num - minVal]++; // Increment the count for the corresponding number.
  }

  int remaining = k; // Initialize the remaining count.
  for (int i = count.length - 1; i >= 0; i--) { // Iterate from the largest number to the smallest.
    remaining -= count[i]; // Subtract the count of the current number from remaining.
    if (remaining <= 0) return i + minVal; // If remaining is 0 or negative, we found the kth largest element.
  }
  return -1; // Return -1 if k is invalid.
}

void main() {
  List<int> nums = [3, 2, 3, 1, 2, 4, 5, 5, 6];
  int k = 4;
  print("Sorting Approach: ${findKthLargestSorting(List.from(nums), k)}");
  print("Min Heap Approach: ${findKthLargestMinHeap(List.from(nums), k)}");
  print("Max Heap Approach: ${findKthLargestMaxHeap(List.from(nums), k)}");
  print("Quickselect Approach: ${quickSelect(List.from(nums), k)}");
  print("Counting Sort Approach: ${findKthLargestCountingSort(List.from(nums), k)}");
}

File: C++/13-Heap(PriorityQueue)/Algorithms/1-MinHeap-MaxHeap/MinHeap-MaxHeap.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <functional> // Required for std::greater and std::less

// Approach 1: Manual Implementation (Max Heap and Min Heap)
//   - Provides the most control and understanding of the heap data structure.
//   - Implements the basic heap operations: insert, getTop, removeTop, and heapify.
//   - Uses a vector as the underlying storage.
//   - Includes both Max Heap and Min Heap implementations in the same struct.

struct ManualHeap {
    std::vector<int> maxHeap; // Vector to store the elements of the max heap
    std::vector<int> minHeap; // Vector to store the elements of the min heap

    // Helper function to maintain the max heap property.
    void maxHeapify(std::vector<int>& heap, int i) {
        int left = 2 * i + 1;  // Index of the left child
        int right = 2 * i + 2; // Index of the right child
        int largest = i;       // Index of the largest element among the node and its children

        if (left < heap.size() && heap[left] > heap[largest]) {
            largest = left;
        }

        if (right < heap.size() && heap[right] > heap[largest]) {
            largest = right;
        }

        if (largest != i) {
            std::swap(heap[i], heap[largest]); // Swap if the largest is not the current node
            maxHeapify(heap, largest);         // Recursively heapify the affected subtree
        }
    }

    // Helper function to maintain the min heap property.
    void minHeapify(std::vector<int>& heap, int i) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        int smallest = i;

        if (left < heap.size() && heap[left] < heap[smallest]) {
            smallest = left;
        }

        if (right < heap.size() && heap[right] < heap[smallest]) {
            smallest = right;
        }

        if (smallest != i) {
            std::swap(heap[i], heap[smallest]);
            minHeapify(heap, smallest);
        }
    }

    // Function to insert an element into the max heap.
    void insertMaxHeap(int value) {
        maxHeap.push_back(value); // Add the new element to the end of the vector
        int i = maxHeap.size() - 1; // Index of the newly inserted element

        // Maintain the max heap property by moving the element up the tree if necessary.
        while (i > 0 && maxHeap[(i - 1) / 2] < maxHeap[i]) {
            std::swap(maxHeap[i], maxHeap[(i - 1) / 2]); // Swap with the parent
            i = (i - 1) / 2;                           // Update the index to the parent's index
        }
    }

    // Function to insert an element into the min heap.
    void insertMinHeap(int value) {
        minHeap.push_back(value);
        int i = minHeap.size() - 1;

        while (i > 0 && minHeap[(i - 1) / 2] > minHeap[i]) {
            std::swap(minHeap[i], minHeap[(i - 1) / 2]);
            i = (i - 1) / 2;
        }
    }

    // Function to get the top element (maximum) of the max heap.
    int getMaxTop() {
        if (maxHeap.empty()) {
            std::cout << "Max Heap is empty" << std::endl;
            return -1; // Return -1 to indicate an empty heap (or throw an exception)
        }
        return maxHeap[0]; // The maximum element is always at the root (index 0)
    }

    // Function to get the top element (minimum) of the min heap.
    int getMinTop() {
        if (minHeap.empty()) {
            std::cout << "Min Heap is empty" << std::endl;
            return -1;
        }
        return minHeap[0];
    }

    // Function to remove the top element (maximum) from the max heap.
    void removeMaxTop() {
        if (maxHeap.empty()) {
            std::cout << "Max Heap is empty, cannot remove top" << std::endl;
            return;
        }
        maxHeap[0] = maxHeap.back(); // Replace the root with the last element
        maxHeap.pop_back();         // Remove the last element
        maxHeapify(maxHeap, 0);       // Restore the max heap property
    }

    // Function to remove the top element (minimum) from the min heap.
    void removeMinTop() {
        if (minHeap.empty()) {
            std::cout << "Min Heap is empty, cannot remove top" << std::endl;
            return;
        }
        minHeap[0] = minHeap.back();
        minHeap.pop_back();
        minHeapify(minHeap, 0);
    }

     // Function to build a max heap from a vector.
    void buildMaxHeap(std::vector<int>& arr) {
        maxHeap = arr;
        for (int i = maxHeap.size() / 2 - 1; i >= 0; i--) {
            maxHeapify(maxHeap, i);
        }
    }

    // Function to build a min heap from a vector.
    void buildMinHeap(std::vector<int>& arr) {
        minHeap = arr;
        for (int i = minHeap.size() / 2 - 1; i >= 0; i--) {
            minHeapify(minHeap, i);
        }
    }
};

// Approach 2: Using std::priority_queue (Max Heap)
//   - Leverages the C++ Standard Library's priority_queue, which is a max heap by default.
//   - Provides a simple and efficient way to use a max heap.
//   - The priority_queue automatically maintains the heap property.

void useStdPriorityQueueMax() {
    std::priority_queue<int> maxHeap; // Declares a max heap of integers

    maxHeap.push(5); // Inserts elements into the heap
    maxHeap.push(2);
    maxHeap.push(8);
    maxHeap.push(1);

    std::cout << "Max Heap using std::priority_queue (Max):" << std::endl;
    while (!maxHeap.empty()) {
        std::cout << maxHeap.top() << " "; // Accesses the top element (maximum)
        maxHeap.pop();                   // Removes the top element
    }
    std::cout << std::endl;
}

// Approach 3: Using std::priority_queue (Min Heap)
//   - Uses std::priority_queue with a custom comparator (std::greater<int>) to create a min heap.
//   - Demonstrates how to customize the behavior of std::priority_queue.

void useStdPriorityQueueMin() {
    // Declares a min heap of integers using std::greater<int> as the comparator
    std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;

    minHeap.push(5);
    minHeap.push(2);
    minHeap.push(8);
    minHeap.push(1);

    std::cout << "Min Heap using std::priority_queue (Min):" << std::endl;
    while (!minHeap.empty()) {
        std::cout << minHeap.top() << " ";
        minHeap.pop();
    }
    std::cout << std::endl;
}

// Approach 4: Using std::make_heap (Max Heap)
//  - Uses `std::make_heap`, `std::push_heap`, and `std::pop_heap` to manage a heap within a vector.
//  - Provides more control over the underlying vector than `std::priority_queue`.

void useStdMakeHeapMax() {
    std::vector<int> heapVector = {5, 2, 8, 1}; // Initialize a vector

    std::make_heap(heapVector.begin(), heapVector.end()); // Convert the vector into a max heap

    std::cout << "Max Heap using std::make_heap (Max):" << std::endl;
    for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    std::cout << "Top element: " << heapVector.front() << std::endl; // Access the top element

    std::pop_heap(heapVector.begin(), heapVector.end()); // Move the top element to the end
    int poppedValue = heapVector.back();             // Get the value that was at the top
    heapVector.pop_back();                           // Remove the element from the end

    std::cout << "Popped element: " << poppedValue << std::endl;

    std::cout << "Heap after pop:" << std::endl;
     for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    heapVector.push_back(6);             // Add a new element to the vector
    std::push_heap(heapVector.begin(), heapVector.end()); // Restore the heap property

     std::cout << "Heap after push:" << std::endl;
     for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;
}

// Approach 5: Using std::make_heap (Min Heap)
//   - Demonstrates using `std::make_heap` with a custom comparator to create a min heap.

void useStdMakeHeapMin() {
    std::vector<int> heapVector = {5, 2, 8, 1};
    // Use std::greater<int> to create a min heap
    std::make_heap(heapVector.begin(), heapVector.end(), std::greater<int>());

    std::cout << "Min Heap using std::make_heap (Min):" << std::endl;
     for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    std::cout << "Top element: " << heapVector.front() << std::endl;

    std::pop_heap(heapVector.begin(), heapVector.end(), std::greater<int>());
    int poppedValue = heapVector.back();
    heapVector.pop_back();

    std::cout << "Popped element: " << poppedValue << std::endl;
    std::cout << "Heap after pop:" << std::endl;
     for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;

    heapVector.push_back(6);
    std::push_heap(heapVector.begin(), heapVector.end(), std::greater<int>());
     std::cout << "Heap after push:" << std::endl;
     for (int value : heapVector) {
        std::cout << value << " ";
    }
    std::cout << std::endl;
}

int main() {
    // Approach 1: Manual Implementation
    std::cout << "Approach 1: Manual Implementation" << std::endl;
    ManualHeap manualHeap;

    // Max Heap
    manualHeap.insertMaxHeap(5);
    manualHeap.insertMaxHeap(2);
    manualHeap.insertMaxHeap(8);
    manualHeap.insertMaxHeap(1);

    std::cout << "Max Heap: ";
    std::vector<int> tempMax = manualHeap.maxHeap; //copy the maxHeap
    while(!tempMax.empty()){
        std::cout<< tempMax[0] << " ";
        tempMax.erase(tempMax.begin());
    }
    std::cout << std::endl;
    std::cout << "Max Top: " << manualHeap.getMaxTop() << std::endl; // 8
    manualHeap.removeMaxTop();
    std::cout << "Max Top after removal: " << manualHeap.getMaxTop() << std::endl; // 5

     std::vector<int> initialMaxHeap = { 10, 5, 3, 7, 1 };
    manualHeap.buildMaxHeap(initialMaxHeap);

    std::cout << "Max Heap after buildMaxHeap: ";
     tempMax = manualHeap.maxHeap; //copy the maxHeap
    while(!tempMax.empty()){
        std::cout<< tempMax[0] << " ";
         tempMax.erase(tempMax.begin());
    }
    std::cout << std::endl;


    // Min Heap
    manualHeap.insertMinHeap(5);
    manualHeap.insertMinHeap(2);
    manualHeap.insertMinHeap(8);
    manualHeap.insertMinHeap(1);

    std::cout << "Min Heap: ";
    std::vector<int> tempMin = manualHeap.minHeap; //copy the minHeap
    while(!tempMin.empty()){
        std::cout<< tempMin[0] << " ";
        tempMin.erase(tempMin.begin());
    }
    std::cout << std::endl;
    std::cout << "Min Top: " << manualHeap.getMinTop() << std::endl; // 1
    manualHeap.removeMinTop();
    std::cout << "Min Top after removal: " << manualHeap.getMinTop() << std::endl; // 2

     std::vector<int> initialMinHeap = { 10, 5, 3, 7, 1 };
    manualHeap.buildMinHeap(initialMinHeap);
    std::cout << "Min Heap after buildMinHeap: ";
    tempMin = manualHeap.minHeap; //copy the minHeap
    while(!tempMin.empty()){
        std::cout<< tempMin[0] << " ";
        tempMin.erase(tempMin.begin());
    }
    std::cout << std::endl;

    // Approach 2: Using std::priority_queue (Max Heap)
    std::cout << "\nApproach 2: Using std::priority_queue (Max Heap)" << std::endl;
    useStdPriorityQueueMax();

    // Approach 3: Using std::priority_queue (Min Heap)
    std::cout << "\nApproach 3: Using std::priority_queue (Min Heap)" << std::endl;
    useStdPriorityQueueMin();

    // Approach 4: Using std::make_heap (Max Heap)
    std::cout << "\nApproach 4: Using std::make_heap (Max Heap)" << std::endl;
    useStdMakeHeapMax();

    // Approach 5: Using std::make_heap (Min Heap)
    std::cout << "\nApproach 5: Using std::make_heap (Min Heap)" << std::endl;
    useStdMakeHeapMin();

    return 0;
}
